faunadb-java/src/main/java/com/faunadb/client/query/Expr.java

```diff
@@ -0,0 +1,230 @@
+package com.faunadb.client.query;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.faunadb.client.types.Ref;
+import com.faunadb.client.types.SetRef;
+import com.faunadb.client.types.Value;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.util.Map;
+
+import static java.lang.String.format;
+
+/**
+ * A query language expression. Constructors for this class are at the {@link Language} class.
+ */
+public abstract class Expr implements Value {
```

Comment:
Let's discuss the details over skype tomorrow, but here are some initial thoughts: 

The big issue is that as far as API consumption goes, an Expr should not be a Value, but the opposite is true (modulo escaping. I agree that implicit conversions would be better, but this is all we have in java). IMO that leads to the need to make that typing relationship fall out, regardless of whatever else needs to be done.

1. I don't care too much about cyclic deps, assuming javac can handle it since it's an implementation detail more than anything else.

2. We should control all of the points where escaping needs to happen, so having it be a runtime error is OK, assuming we have tests covering callsites.

3. I'm not sure expressions should have a generic `value`. However if expressions have an `escapedValue` or `wireFormat` or something, that could be an appropriate place to perform escaping.